spec/services/webhooks/base_service_spec.rb

```diff
@@ -0,0 +1,202 @@
+        expect(webhook.webhook_type).to eq('dummy.test')
+        expect(webhook.endpoint).to eq(organization.webhook_url)
+        expect(webhook.object_id).to eq(invoice.id)
+        expect(webhook.object_type).to eq('Invoice')
+        expect(webhook.http_status).to eq(200)
+        expect(webhook.response).to eq('Success')
+      end
+    end
+
+    context 'with a previous failed webhook' do
+      let(:previous_webhook) { create(:webhook, :failed, organization:) }
+
+      it 'succeeds the retried webhook' do
+        webhook_service.call
+
+        previous_webhook.reload
+
+        aggregate_failures do
+          expect(previous_webhook).to be_succeeded
+          expect(previous_webhook.http_status).to eq(200)
+          expect(previous_webhook.retries).to eq(1)
+          expect(previous_webhook.last_retried_at).not_to be_nil
+        end
+      end
+    end
+
+    context 'without webhook_url' do
+      let(:webhook_url) { nil }
+
+      it 'does not call the organization webhook url' do
+        webhook_service.call
+
+        expect(LagoHttpClient::Client).not_to have_received(:new)
+        expect(lago_client).not_to have_received(:post)
+      end
+    end
+
+    context 'when client returns an error' do
+      let(:error_body) do
+        {
+          message: 'forbidden',
+        }
+      end
+
+      before do
+        allow(LagoHttpClient::Client).to receive(:new)
+          .with(organization.webhook_url)
+          .and_return(lago_client)
+        allow(lago_client).to receive(:post)
+          .and_raise(
+            LagoHttpClient::HttpError.new(403, error_body.to_json, ''),
+          )
+      end
+
+      it 'creates a failed webhook' do
+        webhook_service.call
+
+        webhook = Webhook.first
+
+        aggregate_failures do
+          expect(webhook).to be_failed
+          expect(webhook.http_status).to eq(403)
+        end
+      end
+
+      it 'enqueues a SendWebhookJob' do
+        expect { webhook_service.call }.to have_enqueued_job(SendWebhookJob)
+      end
+
+      context 'with a previous failed webhook' do
+        let(:previous_webhook) { create(:webhook, :failed, organization:) }
+
+        it 'fails the retried webhooks' do
+          webhook_service.call
+
+          previous_webhook.reload
+
+          aggregate_failures do
+            expect(previous_webhook).to be_failed
+            expect(previous_webhook.http_status).to eq(403)
+            expect(previous_webhook.retries).to eq(1)
+            expect(previous_webhook.last_retried_at).not_to be_nil
+          end
+        end
+
+        context 'when the previous failed webhook have been retried 3 times' do
+          let(:previous_webhook) { create(:webhook, :failed, organization:, retries: 2) }
+
+          it 'does not enqueue a SendWebhookJob' do
+            expect { webhook_service.call }.not_to have_enqueued_job(SendWebhookJob)
+          end
+        end
+      end
+    end
+  end
+
+  describe '.generate_headers' do
```

Comment:
Note for later: it feels a bit weird to test private method (I know it was like that in before).
To fix this, maybe we should think at exporting the signature logic into a proper service with its own test suite